substrate/runtime-support/src/storage/generator.rs

```diff
@@ -114,225 +146,119 @@ pub trait StorageList<T: codec::Slicable> {
 
 /// A strongly-typed map in storage.
 pub trait StorageMap<K: codec::Slicable, V: codec::Slicable> {
+	/// The type that get/take returns.
+	type Query;
+
 	/// Get the prefix key in storage.
 	fn prefix() -> &'static [u8];
 
 	/// Get the storage key used to fetch a value corresponding to a specific key.
 	fn key_for(x: &K) -> Vec<u8>;
 
+	/// true if the value is defined in storage.
+	fn exists<S: Storage>(key: &K, storage: &S) -> bool {
+		storage.exists(&Self::key_for(key)[..])
+	}
+
 	/// Load the value associated with the given key from the map.
-	fn get<S: Storage>(key: &K, storage: &S) -> Option<V>;
+	fn get<S: Storage>(key: &K, storage: &S) -> Self::Query;
+
+	/// Take the value under a key.
+	fn take<S: Storage>(key: &K, storage: &S) -> Self::Query;
 
 	/// Store a value to be associated with the given key from the map.
-	fn insert<S: Storage>(key: &K, val: &V, storage: &S);
+	fn insert<S: Storage>(key: &K, val: &V, storage: &S) {
+		storage.put(&Self::key_for(key)[..], val);
+	}
 
 	/// Remove the value under a key.
-	fn remove<S: Storage>(key: &K, storage: &S);
-
-	/// Take the value under a key.
-	fn take<S: Storage>(key: &K, storage: &S) -> Option<V>;
+	fn remove<S: Storage>(key: &K, storage: &S) {
+		storage.kill(&Self::key_for(key)[..]);
+	}
 }
 
 #[macro_export]
 #[doc(hidden)]
 macro_rules! __storage_items_internal {
 	// generator for values.
-	(($($vis:tt)*) $name: ident: $key: expr => $ty:ty) => {
+	(($($vis:tt)*) ($get_fn:ident) ($gettype:ty) ($getter:ident) ($taker:ident) $name:ident : $key:expr => $ty:ty) => {
+		__storage_items_internal!{ ($($vis)*) () ($gettype) ($getter) ($taker) $name : $key => $ty }
+		pub fn $get_fn() -> $gettype { <$name as $crate::storage::generator::StorageValue<$ty>> :: get(&$crate::storage::RuntimeStorage) }
```

Comment:
tightly coupling the macro to `RuntimeStorage` (which is useless outside of runtime context) isn't helpful.

We have a `$name::get()` function anyway via the parent `storage` module.

Reply:
yeah but it's crappy to have half of the getters use the idiomatic `lower_case_function()` and half using `PascalCaseName::get()`. you need to guess which is which.

Reply:
if there's a better way of doing this, then happy to go with it, but i don't see it.

Reply:
fine for now, although I dislike the coupling.